### PR TITLE
Let production build environment assertion allow using Java 8 to run the build

### PR DESCRIPTION
**Pull request by eskatos**
use Java 8 to run the build
and Java 7 for compilation
and Java 8 for running tests

### Context
As part of [dogfooding the Gradle Kotlin DSL](https://github.com/gradle/kotlin-dsl/issues/592), we need to run the `gradle/gradle` build with Java >=8. This change is a follow up for #3735.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -291,9 +294,29 @@ task patchExternalModules(type: PatchExternalModules) {
 
 task verifyIsProductionBuildEnvironment {
     doLast {
-        assert javaVersion.java7: "Must use a Java 7 compatible JVM to perform this build. Current JVM is ${jvm}"
+        assert javaVersion.java8:
+            "Must use a Java 8 compatible JVM to perform this build. Current JVM is ${jvm}"
+
+        def javaHomeProperty = { propertyName -> project.findProperty(propertyName) ?: System.getProperty(propertyName) }
```
Comment (wolfs) [inline build.gradle]: ❌ Note that the name of the property (`java7Home`) and the name of the system property (`java7.home`) is different for the compilation property.
I guess we should make sure that the project properties are set at this point.
https://github.com/gradle/gradle/blob/fd3486cde0040289e81903d0ec4971b82e64f0d4/gradle/compile.gradle#L5-L11

Comment (eskatos) [reply build.gradle]: Addressed in https://github.com/gradle/gradle/pull/3747

**Comment by eskatos** [comment]
> Can we merge this or will the distribution build now fail?

**Comment by eskatos** [comment]
> Can we merge this or will the distribution build now fail?

Right, we need to get https://github.com/gradle/gradle/pull/3749 and https://github.com/gradle/gradle/pull/3750 in first.

**Comment by wolfs** [comment]
I guess https://github.com/gradle/gradle/pull/3750 doesn't matter so much.
We basically need to merge this PR together with https://github.com/gradle/gradle/pull/3749, if not the distribution build will fail.

**Pull request merged**